File: mobile/modules/glasses-media/ios/GlassesMediaRelayModule.swift
import ExpoModulesCore
import Foundation

public final class GlassesMediaRelayModule: Module {
    private let queue = DispatchQueue(label: "com.mentra.glassesmedia.relay")
    private var session: ManagedRelaySession?

    public func definition() -> ModuleDefinition {
        Name("MentraGlassesMediaRelay")
        Events("onRelayState")

        AsyncFunction("prepare") { (options: [String: Any], promise: Promise) in
            guard let id = options["attemptId"] as? String,
                  let ingest = options["ingestUrl"] as? String, let url = URL(string: ingest), url.scheme == "https",
                  let ssid = options["ssid"] as? String, let password = options["password"] as? String
            else {
                throw LocalMediaError("Invalid relay options")
            }
            self.queue.async {
                guard self.session == nil else { promise.reject(LocalMediaError("Previous relay has not stopped")); return }
                let session = ManagedRelaySession(id: id, queue: self.queue)
                self.session = session
                session.start(ssid: ssid, password: password, endpoint: url,
                              captureAudio: options["captureAudio"] as? Bool ?? true,
                              bitrate: options["bitrate"] as? Int ?? 2_000_000,
                              onState: { [weak self] state, reason in
                                  self?.sendEvent("onRelayState", ["attemptId": id, "state": state, "reason": reason])
                              }) { result in
                    switch result {
                    case let .success(url): promise.resolve(url)
                    case let .failure(error): promise.reject(error)
                    }
                }
            }
        }

        AsyncFunction("stop") { (id: String, promise: Promise) in
            self.queue.async {
                guard let session = self.session, session.id == id else { promise.resolve(nil); return }
                session.stop {
                    if self.session === session { self.session = nil }
                    promise.resolve(nil)
                }
            }
        }

        OnDestroy {
            self.queue.async {
                self.session?.stop { self.session = nil }
            }
        }
    }
}

/// Serialized owner of both native peers and the persistent hotspot join; no JS media callbacks.
private final class ManagedRelaySession {
    let id: String
    private let queue: DispatchQueue
    private let hotspot = GlassesHotspotNetwork()
    private var receiver: LocalWhipIngestSource?
    private var publisher: PhoneWhipPublisher?
    private var stopped = false
    private var stopFinished = false
    private var stopReplies: [() -> Void] = []
    private var ready: ((Result<String, Error>) -> Void)?

    init(id: String, queue: DispatchQueue) {
        self.id = id; self.queue = queue
    }

    func start(ssid: String, password: String, endpoint: URL, captureAudio: Bool, bitrate: Int,
               onState: @escaping (String, String) -> Void, completion: @escaping (Result<String, Error>) -> Void)
    {
        ready = completion
        hotspot.onLost = { reason in onState("failed", reason) }
        hotspot.join(ssid: ssid, passphrase: password) { result in
            self.queue.async {
                guard !self.stopped else { return }
                switch result {
                case let .failure(error): self.finish(.failure(error))
                case let .success(address):
                    self.hotspot.awaitInternet { usable, _ in
                        self.queue.async {
                            guard !self.stopped else { return }
                            guard usable else { self.finish(.failure(LocalMediaError("Turn on phone mobile data to stream through the glasses hotspot"))); return }
                            let publisher = PhoneWhipPublisher(endpoint: endpoint, captureAudio: captureAudio, bitrate: bitrate, onState: onState)
                            self.publisher = publisher
                            let receiver = LocalWhipIngestSource()
                            self.receiver = receiver
                            receiver.onFrame = { [weak publisher] in publisher?.onFrame($0) }
                            receiver.onPcm = { [weak publisher] in publisher?.onPcm($0, rate: $1, channels: $2) }
                            receiver.onStateChange = { state, reason in if state == .failed { onState("failed", "Glasses receiver: \(reason)") } }
                            receiver.setPcmDeliveryEnabled(captureAudio)
                            receiver.prepare(config: SourceConfig(url: "", kind: .softap, bindAddress: address)) { result in
                                self.queue.async {
                                    guard !self.stopped else { return }
                                    if case .success = result { publisher.start() }
                                    self.finish(result)
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    private func finish(_ result: Result<String, Error>) {
        let callback = ready; ready = nil; callback?(result)
    }

    func stop(completion: @escaping () -> Void) {
        if stopFinished { completion(); return }
        stopReplies.append(completion)
        guard !stopped else { return }
        stopped = true
        finish(.failure(LocalMediaError("Relay cancelled")))
        hotspot.onLost = nil
        let group = DispatchGroup()
        if let receiver {
            group.enter(); receiver.stop { group.leave() }
        }
        if let publisher {
            group.enter(); publisher.stop { group.leave() }
        }
        group.enter(); hotspot.leave { group.leave() }
        group.notify(queue: queue) {
            self.receiver = nil; self.publisher = nil; self.stopFinished = true
            let replies = self.stopReplies; self.stopReplies.removeAll(); replies.forEach { $0() }
        }
    }
}

File: mobile/modules/glasses-media/android/src/main/java/com/mentra/glassesmedia/GlassesMediaRelayModule.kt
package com.mentra.glassesmedia

import android.content.Intent
import android.os.Handler
import android.os.Looper
import android.os.PowerManager
import android.provider.Settings
import com.mentra.glassesmedia.network.*
import com.mentra.glassesmedia.publisher.PhoneWhipPublisher
import com.mentra.glassesmedia.source.*
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.concurrent.Executors

/** Native resources for one coordinator-owned relay attempt. All transitions use one executor. */
class GlassesMediaRelayModule : Module() {
  private val worker = Executors.newSingleThreadExecutor()
  private var activeId: String? = null
  private var network: ScopedSoftApNetwork? = null
  private var internet: InternetHold? = null
  private var source: LocalWhipIngestSource? = null
  private var publisher: PhoneWhipPublisher? = null
  private var wakeLock: PowerManager.WakeLock? = null

  override fun definition() = ModuleDefinition {
    Name("MentraGlassesMediaRelay")
    Events("onRelayState")

    AsyncFunction("prepare") { options: Map<String, Any?>, promise: Promise ->
      worker.execute {
        try {
          check(activeId == null) { "Previous relay has not stopped" }
          val id = requireNotNull(options["attemptId"] as? String)
          val endpoint = requireNotNull(options["ingestUrl"] as? String)
          val ssid = requireNotNull(options["ssid"] as? String)
          val password = requireNotNull(options["password"] as? String)
          val context = requireNotNull(appContext.reactContext).applicationContext
          activeId = id
          val hold = InternetHold(context).also { internet = it }
          check(hold.awaitValidatedCellular().validated) { "Turn on phone mobile data to stream through the glasses hotspot" }
          val scoped = ScopedSoftApNetwork(context).also { network = it }
          if (!scoped.isWifiEnabled()) {
            Handler(Looper.getMainLooper()).post {
              appContext.currentActivity?.startActivity(Intent(Settings.Panel.ACTION_WIFI))
            }
            check(scoped.awaitWifiEnabled()) { "Phone Wi-Fi is disabled" }
            Thread.sleep(ScopedSoftApNetwork.WIFI_ENABLE_SETTLE_MS)
          }
          scoped.join(ssid, password, object : ScopedSoftApNetwork.Listener {
            override fun onAvailable(network: android.net.Network, localIpv4: String) = Unit
            override fun onLost(error: ScopedNetworkError) { emit(id, "failed", "Glasses hotspot connection was lost") }
          })
          ScopedNetworkChangeDetector.setRelayNetwork(scoped)
          val outgoing = PhoneWhipPublisher(context, endpoint, options["captureAudio"] != false,
            (options["bitrate"] as? Number)?.toInt() ?: 2_000_000) { state, reason -> emit(id, state, reason) }
          publisher = outgoing
          val incoming = LocalWhipIngestSource(context, VideoFrameListener(outgoing::onVideoFrame),
            PcmListener(outgoing::onPcm), scopedNetwork = scoped)
          source = incoming
          incoming.setStateListener { state, reason ->
            if (state == SourceState.FAILED) emit(id, "failed", "Glasses receiver: $reason")
          }
          incoming.setPcmDeliveryEnabled(options["captureAudio"] != false)
          incoming.start(SourceConfig("", SourceKind.SOFTAP, scoped.localIpv4()))
          wakeLock = context.getSystemService(PowerManager::class.java)
            .newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "Mentra:managed-relay").apply { acquire() }
          outgoing.start()
          promise.resolve(requireNotNull(incoming.ingestUrl))
        } catch (error: Exception) {
          // The owner still calls stop: preserve ownership until every acquired resource is gone.
          promise.reject("RELAY_PREPARE_FAILED", error.message, error)
        }
      }
    }

    AsyncFunction("stop") { attemptId: String, promise: Promise ->
      worker.execute {
        try {
          if (activeId == attemptId) cleanup()
          promise.resolve(null)
        } catch (error: Exception) { promise.reject("RELAY_STOP_FAILED", error.message, error) }
      }
    }

    OnDestroy { worker.execute { runCatching { cleanup() }; worker.shutdown() } }
  }

  private fun emit(id: String, state: String, reason: String) {
    sendEvent("onRelayState", mapOf("attemptId" to id, "state" to state, "reason" to reason))
  }

  private fun cleanup() {
    source?.setStateListener(null)
    // Try every cleanup even if one fails. Keep the slot occupied on any failure.
    var failure: Exception? = null
    fun step(action: () -> Unit) { try { action() } catch (e: Exception) { failure = failure ?: e } }
    step { source?.stop(); source = null }
    step { publisher?.close(); publisher = null }
    step { network?.release(); network = null; ScopedNetworkChangeDetector.setRelayNetwork(null) }
    step { internet?.close(); internet = null }
    step { wakeLock?.let { if (it.isHeld) it.release() }; wakeLock = null }
    failure?.let { throw it }
    activeId = null
  }
}
